### PR TITLE
Sanitizing parameters in Audit Logs

### PR DESCRIPTION
**Pull request by lashinijay**
### Proposed changes in this pull request
Please note the $subject.

### CODE REVIEW

**Comment by lashinijay** [comment]
The framework version needs to be bumped [here](https://github.com/wso2-extensions/identity-data-publisher-authentication/pull/104/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L232) after [https://github.com/wso2/carbon-identity-framework/pull/4296](https://github.com/wso2/carbon-identity-framework/pull/4296) merged.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 1]
```diff
@@ -179,18 +182,18 @@ protected void doPublishAuthenticationFailure(AuthenticationAuditData authentica
...
         AUDIT_LOG.info(String.format(
                 FrameworkConstants.AUDIT_MESSAGE,
-                authenticationData.getAuthenticatedUser(),
+                LoggerUtils.maskContent(authenticationData.getAuthenticatedUser()),
                 "Login",
                 "ApplicationAuthenticationFramework", auditData, FrameworkConstants.AUDIT_FAILED));
     }
 
-    protected void doPublishSessionTermination(AuthenticationContext context, String username,
+    protected void doPublishSessionTermination(AuthenticationContext context, String username, String userId,
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: shouldn't change the method signature of protected methods

Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: fixed

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java [hunk 2]
```diff
@@ -80,7 +81,11 @@ public static AuthenticationAuditData createAuthenticationAudiDataObject(Event e
             authenticationAuditData.setAuthenticatedIdps(getIdentityProviderList(context, status));
 
         }
-
+        if (StringUtils.isNotBlank(authenticationAuditData.getTenantDomain())){
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]:
```suggestion
        if (StringUtils.isNotBlank(authenticationAuditData.getTenantDomain())) {
```

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java [hunk 3]
```diff
@@ -80,7 +81,11 @@ public static AuthenticationAuditData createAuthenticationAudiDataObject(Event e
             authenticationAuditData.setAuthenticatedIdps(getIdentityProviderList(context, status));
 
         }
-
+        if (StringUtils.isNotBlank(authenticationAuditData.getTenantDomain())) {
+            String userId = IdentityUtil.getInitiatorId(getUserNameForAuthenticationStep(params),
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]: check how AuthenticatedUser is resolved in https://github.com/wso2-extensions/identity-data-publisher-authentication/pull/104/files#diff-7e633e100b942bead2770c632a75a52a7f96ee94ae495a277f5c33e043998fa7R67-R82
better to have the userid resolved in those places

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 4]
```diff
@@ -179,18 +183,19 @@ protected void doPublishAuthenticationFailure(AuthenticationAuditData authentica
...
-                authenticationData.getAuthenticatedUser(),
+                LoggerUtils.maskContent(authenticationData.getAuthenticatedUser()),
                 "Login",
                 "ApplicationAuthenticationFramework", auditData, FrameworkConstants.AUDIT_FAILED));
     }
 
-    protected void doPublishSessionTermination(AuthenticationContext context, String username,
-                                               String tenantDomain, String authenticatedIDPs) {
+    protected void doPublishSessionTermination(AuthenticationContext context, String username, String tenantDomain,
+                                               String authenticatedIDPs) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: let's revert this change

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 5]
```diff
@@ -123,7 +127,7 @@ protected void doPublishAuthenticationStepSuccess(AuthenticationAuditData authen
 
         AUDIT_LOG.info(String.format(
                 FrameworkConstants.AUDIT_MESSAGE,
-                authenticationData.getAuthenticatedUser(),
+                maskedUsername,
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: is the user id always not available in this case?

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml [hunk 6]
```diff
@@ -38,6 +38,10 @@
             <groupId>org.wso2.carbon</groupId>
             <artifactId>org.wso2.carbon.utils</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.user.core</artifactId>
+        </dependency>
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml]: why do we need this?

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml [hunk 7]
```diff
@@ -102,8 +110,10 @@
                             org.osgi.framework; version="${osgi.framework.imp.pkg.version.range}",
                             org.osgi.service.component.*; version="${osgi.service.component.imp.pkg.version.range}",
                             org.wso2.carbon; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.user.core.util.*; version="${carbon.kernel.package.import.version.range}",
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml]: is this needed?

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml (hunk 6 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml:44]: getInitatorId(String username, String tenantDomain) was moved into IdentityUtils as suggested. That's why this dependency is needed.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml (hunk 7 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/pom.xml:113]: getInitatorId(String username, String tenantDomain) was moved into IdentityUtils as suggested. That's why this dependency is needed.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java (hunk 5 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: Yes, it is not available in this scenario.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java (hunk 3 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]: resolved

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java (hunk 4 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: resolved

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java (hunk 3 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]: resolved

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java [hunk 8]
```diff
@@ -63,24 +65,37 @@ public static AuthenticationAuditData createAuthenticationAudiDataObject(Event e
...
         } else if (AuthenticationAuditLoggerConstants.AUDIT_AUTHENTICATION.equals(authType)) {
             if (isUsernameEnabled) {
-                authenticationAuditData.setAuthenticatedUser(getAuthenticatedUserName(context, status));
+                authenticatedUser = getAuthenticatedUserName(context, status);
+                username = authenticatedUser;
+                authenticationAuditData.setAuthenticatedUser(authenticatedUser);
             } else {
-                authenticationAuditData.setAuthenticatedUser(getSubjectIdentifier(context, status));
+                authenticatedUser = getSubjectIdentifier(context, status);
+                username = MultitenantUtils.getTenantAwareUsername(authenticatedUser);
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]: why is tenant aware username obtained here and not in previous conditions?

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 9]
```diff
@@ -123,10 +127,9 @@ protected void doPublishAuthenticationStepSuccess(AuthenticationAuditData authen
 
         AUDIT_LOG.info(String.format(
                 FrameworkConstants.AUDIT_MESSAGE,
-                authenticationData.getAuthenticatedUser(),
+                username,
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:133]: shouldn't it be the user id for the initiator?

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 10]
```diff
@@ -141,7 +144,7 @@ protected void doPublishAuthenticationStepFailure(AuthenticationAuditData authen
 
         AUDIT_LOG.info(String.format(
                 FrameworkConstants.AUDIT_MESSAGE,
-                authenticationData.getAuthenticatedUser(),
+                getUsernameForAuditLog(authenticationData.getAuthenticatedUser()),
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:154]: is the user id not available here? if so better to add a comment

File: components/org.wso2.carbon.identity.data.publisher.authentication.analytics.login/src/test/java/org/wso2/carbon/identity/data/publisher/authentication/analytics/login/AnalyticsLoginDataPublisherUtilsTest.java [hunk 11]
```diff
@@ -62,6 +64,7 @@ public Object[][] getEvent() {
         Map<String, Object> param = new HashMap<>();
         Object userObj = new AuthenticatedUser();
         ((AuthenticatedUser) userObj).setTenantDomain(TENANT_DOMAIN);
+        ((AuthenticatedUser) userObj).setUserId(USER_ID);
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.analytics.login/src/test/java/org/wso2/carbon/identity/data/publisher/authentication/analytics/login/AnalyticsLoginDataPublisherUtilsTest.java]: why is this needed in the test?

Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.analytics.login/src/test/java/org/wso2/carbon/identity/data/publisher/authentication/analytics/login/AnalyticsLoginDataPublisherUtilsTest.java]: Otherwise this test is failing due to a NPE.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java (hunk 9 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:133]: This log is created before the user gets authenticated. Therefore userId is not present in this case.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java (hunk 10 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:154]: Yes, userId is not available as the user is not authenticated yet.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java (hunk 8 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]: In this case, authenticatedUser is resolved as username@tenantDomain format. To derive the username part from it used the getTenantAwareUsername method here. In the previous case getAuthenticatedUserName() method returns the username only. Therefore getTenantAwareUsername was not needed there.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 12]
```diff
@@ -142,6 +145,10 @@ protected void doPublishAuthenticationStepFailure(AuthenticationAuditData authen
                 + "\"";
         auditData = addContextualInfo(auditData, authenticationData);
 
+        /*
+        Here username is used for the initiator instead of userId as userId is not resolved here due to the
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]: better to mention that is when the config is enabled

File: components/org.wso2.carbon.identity.data.publisher.authentication.analytics.login/src/test/java/org/wso2/carbon/identity/data/publisher/authentication/analytics/login/AnalyticsLoginDataPublisherUtilsTest.java (hunk 11 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.analytics.login/src/test/java/org/wso2/carbon/identity/data/publisher/authentication/analytics/login/AnalyticsLoginDataPublisherUtilsTest.java]: the underlying cause was due to the pax-logger. fixed it.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java [hunk 13]
```diff
@@ -63,24 +65,37 @@ public static AuthenticationAuditData createAuthenticationAudiDataObject(Event e
...
-            authenticationAuditData.setTenantDomain(getTenantDomainForAuthentication(context, params, status));
+            tenantDomain = getTenantDomainForAuthentication(context, params, status);
+            authenticationAuditData.setTenantDomain(tenantDomain);
             authenticationAuditData.setStepNo(getStepNoForAuthentication(context, status));
             authenticationAuditData.setAuthenticatedIdps(getIdentityProviderList(context, status));
-
         }
-
+        if (StringUtils.isNotBlank(tenantDomain) && StringUtils.isNotBlank(username)) {
+            String userId = IdentityUtil.getInitiatorId(username, tenantDomain);
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java:103]: what happens if by any chance the username here is not a tenant aware username?

Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java:103]: userId will be null in that case. I am validating whether the userId is null before printing the log. If userId is null then the username will be used as the initiator.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 14]
```diff
@@ -111,22 +114,25 @@ private boolean isAuditLoggerUserNameEnabled(Event event) throws IdentityEventEx
...
                 + "\",\"" + "ServiceProviderName" + "\" : \"" + authenticationData.getServiceProvider()
                 + "\",\"" + "RequestType" + "\" : \"" + authenticationData.getInboundProtocol()
                 + "\",\"" + "RelyingParty" + "\" : \"" + authenticationData.getRelyingParty()
                 + "\",\"" + "AuthenticatedIdP" + "\" : \"" + authenticationData.getAuthenticatedIdps()
                 + "\"";
         auditData = addContextualInfo(auditData, authenticationData);
-
+        /*
+        Here username is used for the initiator instead of userId when the log masking is enabled also, as userId is
+        not resolved yet.User is not authenticated yet.
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:129]: Here the userId cannot be resolved as the user is not authenticated yet.  Hence the username is used for the initiator even when the log masking is enabled.

Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java]:
```suggestion
        Here the userId cannot be resolved as the user is not authenticated yet. Hence the username is used for the initiator even when the log masking is enabled.
```

**Review by dewniMW — DISMISSED** [review]

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java [hunk 15]
```diff
@@ -271,4 +281,61 @@ private String addRemoteAddressAndAgent(String data) {
...
+
+    /**
+     * Returns initiator for audit logs based on log masking config.
+     *
+     * @param username     Username.
+     * @param tenantDomain Tenant domain.
+     * @return initiator. Returns userId if log masking is enabled, if userId cannot be resolved then returns the masked
+     * username.
+     * */
+    private String getInitiator(String username, String tenantDomain) {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:307]: let's return an Optional here

File: pom.xml [hunk 16]
```diff
@@ -115,7 +124,7 @@
             </dependency>
             <!-- Pax Logging -->
             <dependency>
-                <groupId>org.ops4j.pax.logging</groupId>
+                <groupId>org.wso2.org.ops4j.pax.logging</groupId>
```
Comment (chamathns) [inline pom.xml:127]: This is a major dependency change that might cause regressions.  Do we need to change this under this scope?

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java (hunk 15 above)
Comment (lashinijay) [reply components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggingHandler.java:307]: This method won't return null as the initiator in any scenario.

File: pom.xml (hunk 16 above)
Comment (lashinijay) [reply pom.xml:127]: After bumping the framework version there is an error in building repo unless we update this dependency.

File: components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java [hunk 17]
```diff
@@ -63,24 +66,44 @@ public static AuthenticationAuditData createAuthenticationAudiDataObject(Event e
...
-                authenticationAuditData.setAuthenticatedUser(getSubjectIdentifier(context, status));
+                authenticatedUser = getSubjectIdentifier(context, status);
+                /*
+                Tenant aware username is required only if masking logs has been enabled. Added this condition to avoid
+                unnecessary method call in the case of masking logs is disabled.
+                */
+                if (LoggerUtils.isLogMaskingEnable) {
+                    username = StringUtils.isNotBlank(authenticatedUser) ?
+                            MultitenantUtils.getTenantAwareUsername(authenticatedUser) : null;
+                }
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.data.publisher.authentication.audit/src/main/java/org/wso2/carbon/identity/data/publisher/authentication/audit/AuthenticationAuditLoggerUtils.java]:
```suggestion
                if (LoggerUtils.isLogMaskingEnable && StringUtils.isNotBlank(authenticatedUser)) {
                    username =  MultitenantUtils.getTenantAwareUsername(authenticatedUser);
                }
```

**Review by dewniMW — APPROVED** [review]
